llpc/lower/llpcSpirvLowerReverseThreadGroup.cpp

```diff
@@ -0,0 +1,176 @@
+/*
+************************************************************************************************************************
+*
+*  Copyright (C) 2019-2022 Advanced Micro Devices, Inc. All rights reserved.
+*
+***********************************************************************************************************************/
+/**
+ ***********************************************************************************************************************
+ * @file  llpcSpirvLowerReverseThreadGroup.cpp
+ * @brief LLPC source file: contains implementation of class Llpc::SpirvLowerReverseThreadGroup.
+ * @details This pass reverses the gl_WorkGroupId, and also reverse gl_GlobalInvocationId, if they are used.
+ ***********************************************************************************************************************
+ */
+#include "llpcSpirvLowerReverseThreadGroup.h"
+#include "SPIRVInternal.h"
+#include "llpcContext.h"
+#include "llpcDebug.h"
+#include "llpcSpirvLower.h"
+#include "llpcSpirvLowerUtil.h"
+#include "lgc/Builder.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/Pass.h"
+#include "llvm/Support/Debug.h"
+#include "llvm/Support/raw_ostream.h"
+
+#define DEBUG_TYPE "llpc-spirv-lower-reverse-thread-group"
+
+using namespace llvm;
+using namespace SPIRV;
+using namespace Llpc;
+
+namespace Llpc {
+
+// =====================================================================================================================
+// Initializes static members.
+char LegacySpirvLowerReverseThreadGroup::ID = 0;
+
+// =====================================================================================================================
+// Pass creator, creates the pass of SPIR-V lowering operations for reversing thread group
+ModulePass *createLegacySpirvLowerReverseThreadGroup() {
+  return new LegacySpirvLowerReverseThreadGroup();
+}
+
+// =====================================================================================================================
+LegacySpirvLowerReverseThreadGroup::LegacySpirvLowerReverseThreadGroup() : ModulePass(ID) {
+}
+
+// =====================================================================================================================
+// Executes this SPIR-V lowering pass on the specified LLVM module.
+//
+// @param [in/out] module : LLVM module to be run on (empty on entry)
+bool LegacySpirvLowerReverseThreadGroup::runOnModule(Module &module) {
+  return Impl.runImpl(module);
+}
+
+// =====================================================================================================================
+// Executes this SPIR-V lowering pass on the specified LLVM module.
+//
+// @param [in/out] module : LLVM module to be run on (empty on entry)
+// @param [in/out] analysisManager : Analysis manager to use for this transformation
+PreservedAnalyses SpirvLowerReverseThreadGroup::run(Module &module, ModuleAnalysisManager &analysisManager) {
+  runImpl(module);
+  return PreservedAnalyses::none();
```

Comment:
If `runImpl` returns false because nothing was changed, this function can return `PreservedAnalysis::all()`.